### ISSUE TITLE
[FSDP2] model weights are not DTensor after forward pass

### ISSUE THREAD

**Issue by qsh-zh** [pytorch/pytorch#144289]
### 🐛 Describe the bug

When using FSDP2, certain parameters are not DTensor before the backward pass, but appear as DTensor after the backward pass. 

The inconsistency surprises users. For example, when I need to copy the Exponential Moving Average (EMA) model weights to the regular model after the regular model has run some forward steps for evaluation, it causes errors as EMA model weights are in DTensor while some parameters are not DTensor  in regular model.

```python
"""
torchrun --standalone --nproc_per_node=2 test.py
"""
import os
import torch
import torch.distributed as dist
from torch.distributed._tensor.api import DTensor
from torch.distributed._composable.fsdp import fully_shard, MixedPrecisionPolicy
from torch.testing._internal.distributed._tensor.common_dtensor import (
    ModelArgs,
    Transformer,
    TransformerBlock,
)
def main():
    torch.manual_seed(42)
    model_args = ModelArgs(
        n_layers=12,
        vocab_size=50304,
        n_heads=32,
        dim=2048,
        max_seq_len=2048,
        dropout_p=0.0,
    )
    model = Transformer(model_args)
    mp_policy = MixedPrecisionPolicy(param_dtype=torch.bfloat16)
    fsdp_cfg = {"mp_policy": mp_policy}
    for module in model.modules():
        if isinstance(module, TransformerBlock):
            fully_shard(module, **fsdp_cfg)
    fully_shard(model, **fsdp_cfg)
    optim = torch.optim.AdamW(model.parameters(), lr=1e-2)
    inp = torch.randint(0, model_args.vocab_size, (8, 1024), device="cuda")
    loss = model(inp).sum()

    for name, param in model.named_parameters():
        if not isinstance(param, DTensor):
            if dist.get_rank() == 0:
                print(f"Before backward Non-DTensor: {name}")

    loss.backward()

    for name, param in model.named_parameters():
        if not isinstance(param, DTensor):
            if dist.get_rank() == 0:
                print(f"After backward Non-DTensor: {name}")

    optim.step()

if __name__ == "__main__":
    dist.init_process_group(backend="nccl")
    gpu_id = int(os.environ["LOCAL_RANK"])
    device = f"cuda:{gpu_id}"
    torch.cuda.set_device(device)
    rank = gpu_id
    main()
    dist.destroy_process_group()

````

output

```shell
Before backward Non-DTensor: tok_embeddings.weight
Before backward Non-DTensor: pos_embeddings.weight
Before backward Non-DTensor: norm.weight
Before backward Non-DTensor: norm.bias
```

Should not all paramters be with DTensor type?

### Versions

```
PyTorch version: 2.5.1+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.30.2
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jul 29 2024, 16:56:48) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-1055-aws-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.6.68
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3
GPU 2: NVIDIA H100 80GB HBM3
GPU 3: NVIDIA H100 80GB HBM3
GPU 4: NVIDIA H100 80GB HBM3
GPU 5: NVIDIA H100 80GB HBM3
GPU 6: NVIDIA H100 80GB HBM3
GPU 7: NVIDIA H100 80GB HBM3

Nvidia driver version: 550.54.15
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.4.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             192
On-line CPU(s) list:                0-191
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 7R13 Processor
CPU family:                         25
Model:                              1
Thread(s) per core:                 2
Core(s) per socket:                 48
Socket(s):                          2
Stepping:                           1
BogoMIPS:                           5299.99
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch topoext perfctr_core invpcid_single ssbd ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr rdpru wbnoinvd arat npt nrip_save vaes vpclmulqdq rdpid
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          3 MiB (96 instances)
L1i cache:                          3 MiB (96 instances)
L2 cache:                           48 MiB (96 instances)
L3 cache:                           384 MiB (12 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-47,96-143
NUMA node1 CPU(s):                  48-95,144-191
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Vulnerable, no microcode
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:           Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] DISTS-pytorch==0.1
[pip3] flake8==7.1.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cudnn-frontend==1.6.0
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] nvidia-pytriton==0.5.11
[pip3] nvtx==0.2.10
[pip3] onnx==1.16.2
[pip3] onnx-graphsurgeon==0.5.2
[pip3] onnxruntime==1.19.2
[pip3] open_clip_torch==2.20.0
[pip3] optree==0.12.1
[pip3] pynvjitlink==0.3.0
[pip3] pytorch-lightning==2.4.0
[pip3] pytorch-ranger==0.1.1
[pip3] pytorch-triton==3.0.0+dedb7bdf3
[pip3] slangtorch==1.2.6
[pip3] torch==2.5.1
[pip3] torch_automated_profiler==1.10.0
[pip3] torch-fidelity==0.3.0
[pip3] torch-optimizer==0.3.0
[pip3] torch_tensorrt==2.5.0a0
[pip3] torchmetrics==1.4.2
[pip3] torchvision==0.20.0a0
[pip3] triton==3.1.0
[pip3] tritonclient==2.50.0
[conda] Could not collect
```

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @kwen2501 @chauhang

**Comment by qsh-zh** [comment]
maybe @awgu

**Comment by awgu** [comment]
@qsh-zh sorry about this. Let me explain the technical reason first.

Each FSDP module (i.e. the module you call `fully_shard()` on) represents one FSDP parameter group. This group can either be in a `SHARDED` or `UNSHARDED` state.
- For `SHARDED`, the sharded parameters are registered to the module. The unsharded parameters (result of all-gather) are _not_ allocated.
- For `UNSHARDED`, the sharded parameters are still allocated, but now the unsharded parameters are allocated and registered to the module.

If the sharded parameters are registered to the module, then `module.weight` returns the sharded `weight`. If the unsharded parameters are registered to the module, the `module.weight` returns the unsharded `weight`.

We see that as long as the unsharded parameters are allocated, then they are also registered to the module (and hence exposed to the user). There is no case (like in FSDP1) where they could be allocated but not registered. This simplifies the design and implementation since we do not need to handle a case where the sharded parameters were registered and modified by the user but we did not re-all-gather unsharded parameters with the updated sharded parameters since the unsharded parameters were already allocated.

What state an FSDP module is in after forward corresponds directly to the `reshard_after_forward: bool` arg. The confusion is that for the _root_ FSDP module, we automatically clamp `reshard_after_forward` to `False` for the user. This is what is leading to your issue.

The reason that we do this is because we do not want to free the parameters and then immediately re-all-gather them in backward. This was the same thing done in FSDP1. In fact, in FSDP1, if you accessed the parameters after forward but before backward (like `module.weight`), you could get a "dangling reference" where the unsharded parameter object is returned but has no storage. With the FSDP2 design, that dangling reference is hidden; i.e. the user cannot access the unsharded parameter if it is freed.

I would mainly recommend that you use the `FSDPModule.reshard()` API if you want to access the sharded parameters after forward. You can simply loop over your modules like:
```
for module in model.modules():
    if isinstance(module, FSDPModule):
        module.reshard()
```
This will ensure, regardless of if the root or if `reshard_after_forward=False`, all FSDP modules will be in the `SHARDED` state and hence have the sharded parameters (`DTensor`s) registered.

**Comment by qsh-zh** [comment]
@awgu  thank your for detailed and amazing explanation! The design looks interesting!

I have a side question about FSDP2. If a user performs in-place modifications to the model weights, it seems FSDP2 can handle this correctly without requiring extensive code changes. Here is an example of a popular [network design with forced weight normalization](https://github.com/NVlabs/edm2/blob/main/training/networks_edm2.py#L101-L108) that normalizes weights during the forward pass:

```python
        w = self.weight.to(torch.float32)
        if self.training:
            with torch.no_grad():
                self.weight.copy_(normalize(w))  # Forced weight normalization
        w = normalize(w)  # Traditional weight normalization
        w = w * (gain / np.sqrt(w[0].numel()))  # Magnitude-preserving scaling
        w = w.to(x.dtype)
        if w.ndim == 2:
            return x @ w.t()
        assert w.ndim == 4
        return torch.nn.functional.conv2d(x, w, padding=(w.shape[-1] // 2,))
```

After the forward pass, will the unsharded weights be normalized and then sharded again as normalized weights correctly like DDP?

Additionally, does the new design in FSDP2 address [the unclear extra memory usage issue](https://github.com/pytorch/pytorch/issues/120003#issuecomment-2495889172)?

**Comment by awgu** [comment]
@qsh-zh I think if you do in-place modifications to the unsharded weights during forward, then those changes will _not_ propagate to the sharded weights, leading to a correctness issue. You should instead normalize the sharded weights such that the unsharded weights that result from all-gather are what you want.

From a design perspective, it is hard/expensive to know/track when the unsharded weights have been modified to then write the result back to the sharded weights. Moreover, there is no guarantee that the user modifies the unsharded weight in the same way across all ranks, which could lead to hard-to-debug numeric issues.

To help your mental model, under the FSDP1/2 designs, the sharded weight is persistent. It is not reconstructed by resharding the unsharded weight. In fact, reconstructing it does not make sense since under the common fp32-sharded-weight + bf16-unsharded-weight setup, we cannot recover the fp32 sharded weight from the bf16 unsharded weight due to the higher precision for the sharded weight.

> Additionally, is the new design in FSDP2 the fix for https://github.com/pytorch/pytorch/issues/120003#issuecomment-2495889172?

Yes, but this is somewhat orthogonal to all of the previous discussion in this particular Github issue. To avoid memory usage issues, we mainly have to redesign how we manage memory and make sure not to call `recordStream` like mentioned briefly in the other comment.

**Comment by qsh-zh** [comment]
@awgu  thank you for your response and explanation.  It is clear to me now.

**Comment by davidmcclure** [comment]
@awgu Thanks for this thread, really helpful!

> I think if you do in-place modifications to the unsharded weights during forward, then those changes will not propagate to the sharded weights, leading to a correctness issue. You should instead normalize the sharded weights such that the unsharded weights that result from all-gather are what you want.

Is it safe to do in-place modifications on the sharded weights after the forward? Eg, for something like swapping in EMA weights. If you do that, is it necessary to do the recursive `reshard()` to ensure that the updated weights are registered to the module for the next forward?

**Comment by weifengpy** [comment]
> Does anyone know how to make FSDP2 work with EMA? Should EMA store DTensor or Tensor?

ema should store dtensor
